sklearn/mixture/gaussianmixture.py

```diff
@@ -0,0 +1,697 @@
+    covars /= X.shape[0]
+    covars.flat[::len(covars) + 1] += reg_covar
+    return covars
+
+
+def estimate_Gaussian_suffstat_Sk_diag(resp, X, nk, xk, reg_covar):
+    """Compute the covariance matrices for the 'diag' case
+
+    Parameters
+    ----------
+    responsibilities : array-like, shape = (n_samples, n_components)
+
+    X : array-like, shape = (n_samples, n_features)
+
+    nk: array-like, shape = (n_components,)
+
+    xk : array-like, shape = (n_components, n_features)
+
+    reg_covar : float
+
+    Returns
+    -------
+    Sk : array, shape = (n_components, n_features)
+    """
+    avg_X2 = np.dot(resp.T, X * X) / nk[:, np.newaxis]
+    avg_means2 = xk ** 2
+    avg_X_means = xk * np.dot(resp.T, X) / nk[:, np.newaxis]
+    return avg_X2 - 2 * avg_X_means + avg_means2 + reg_covar
+
+
+def estimate_Gaussian_suffstat_Sk_spherical(resp, X, nk, xk, reg_covar):
+    """Compute the covariance matrices for the 'spherical' case
+
+    Parameters
+    ----------
+    responsibilities : array-like, shape = (n_samples, n_components)
+
+    X : array-like, shape = (n_samples, n_features)
+
+    nk: array-like, shape = (n_components,)
+
+    xk : array-like, shape = (n_components, n_features)
+
+    reg_covar : float
+
+    Returns
+    -------
+    Sk : array, shape = (n_components,)
+    """
+    covars = estimate_Gaussian_suffstat_Sk_diag(resp, X, nk, xk, reg_covar)
+    return covars.mean(axis=1)
+
+
+def estimate_Gaussian_suffstat_Sk(resp, X, nk, xk, reg_covar, covar_type):
+    """Compute the covariance matrices
+
+    Parameters
+    ----------
+    resp : array-like, shape = (n_samples, n_components)
+
+    X : array-like, shape = (n_samples, n_features)
+
+    nk: array-like, shape = (n_components,)
+
+    xk : array-like, shape = (n_components, n_features)
+
+    reg_covar : float
+
+    covar_type : string
+
+    Returns
+    -------
+    Sk : array,
+        full : shape = (n_components, n_features, n_features)
+        tied : shape = (n_components, n_features)
+        diag : shape = (n_components, n_features)
+        spherical : shape = (n_components,)
+    """
+    # TODO try methods in sklearn.covariance
+    suffstat_sk_functions = {
+        "full": estimate_Gaussian_suffstat_Sk_full,
+        "tied": estimate_Gaussian_suffstat_Sk_tied,
+        "diag": estimate_Gaussian_suffstat_Sk_diag,
+        "spherical": estimate_Gaussian_suffstat_Sk_spherical}
+    return suffstat_sk_functions[covar_type](resp, X, nk, xk, reg_covar)
+
+
+def estimate_Gaussian_suffstat_xk(resp, X, nk):
+    """Compute the sufficient statistics for Gaussian distribution
+
+    Parameters
+    ----------
+    responsibilities : array-like, shape = (n_samples, n_components)
+
+    X : array-like, shape = (n_samples, n_features)
+
+    covariance_type : string
+
+    Returns
+    -------
+    xk : array-like, shape = (n_components, n_features)
+    """
+    # remove + 10 * EPS
+    return np.dot(resp.T, X) / nk[:, np.newaxis]
```

Comment:
The body of this function could be collapsed in the `_estimate_suffstat` function to skip a useless indirection when reading the code.